### PR TITLE
Add post about django-allow-cidr and ELB health checks

### PR DESCRIPTION
**Pull request by pmac**
Will probably need a thorough check for accuracy for k8s and AWS info.

### CODE REVIEW

File: _posts/2018-02-27-django-k8s-elb-health-checks.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+layout: post
+title: Django, K8s, and ELB Health checks
+author: Paul McLanahan
+excerpt_separator: <!--more-->
+---
+
+As you may have seen in several of our
+[SRE](https://mozilla.github.io/meao/2018/02/16/sre-status/)
+[status](https://mozilla.github.io/meao/2018/01/23/sre-status/)
+[reports](https://mozilla.github.io/meao/2017/12/05/sre-status/)
```
Comment (metadave) [inline _posts/2018-02-27-django-k8s-elb-health-checks.md]: probably needs a comma

File: _posts/2018-02-27-django-k8s-elb-health-checks.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+## AWS ELB Health Checks
+
+Here's the thing: Amazon Web Services' (AWS) Elastic Load Balancers (ELB) do not have many options for
+their health checks. These checks ensure that your app on a particular node in your cluster
+is working as expected. If the check fails the ELB will remove the node from the list of nodes
+to which it will route requests for your app. However, because it's hitting the nodes directly
+it doesn't rely on DNS and directly requests the IP address and port, and it doesn't allow you
+to specify custom headers (e.g. the `Host` header). It also can't do HTTPS because we terminate
+TLS connections at the ELB, so the app nodes speak only plain HTTP back to the ELB. All of that
+means that our health check endpoint needs to do two unique things: allow HTTP connection and
```
Comment (metadave) [inline _posts/2018-02-27-django-k8s-elb-health-checks.md]: should this be `allow HTTP connections` (plural)?

File: _posts/2018-02-27-django-k8s-elb-health-checks.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+allow the IP address that the ELB requests as a valid `Host` header. The first bit is easy enough
+when using Django's in-built `SecurityMiddleware` since it supports the `SECURE_REDIRECT_EXEMPT`
+setting. It's this second requirement that gets more interesting when combined with k8s.
+
+## K8s Routing
+
+The way I understand it (and I'm admittedly no expert) is that k8s (at least the way we use it)
+sets up a "node port" per app (or namespace). To hit that app you can hit any node in the cluster
```
Comment (metadave) [inline _posts/2018-02-27-django-k8s-elb-health-checks.md]: `NodePort` is how it's referred to in the [K8s docs](https://kubernetes.io/docs/concepts/services-networking/service/)

File: _posts/2018-02-27-django-k8s-elb-health-checks.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+at that port and that node will route you to one of the nodes that is running a pod (ususally one
```
Comment (metadave) [inline _posts/2018-02-27-django-k8s-elb-health-checks.md]: typo: `ususally`

File: _posts/2018-02-27-django-k8s-elb-health-checks.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+node that was initially hit, but not for the node that serves the request. This means that we can't
+simply add the IP of the host to the `ALLOWED_HOSTS` list when the app starts. We could get more info
+from AWS' metadata service endpoint, but for security reasons we block that service from all of our
+nodes.
+
+So, the approach could then be to simply add all of the IPs for all of the nodes in the cluster to the
+`ALLOWED_HOSTS` setting and call it done. The problem with this happens when there is a scaling event.
+When a node is killed and a new one started, or the cluster is scaled to include more nodes, you'd need
+to have a way to inform every running pod of this change so they could get the new list of IPs. If they
+didn't then the new node(s) could be immediately excluded from the cluster because health checks would
```
Comment (metadave) [inline _posts/2018-02-27-django-k8s-elb-health-checks.md]: command between `didn't` and `then`

File: _posts/2018-02-27-django-k8s-elb-health-checks.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+return 400s as their IP would not be allowed by Django.
+
+## Enter django-allow-cidr
+
+The way we decided to solve this was by implementing a Django middleware that would allow a range of IP
+addresses defined by a [CIDR][] (Classless Inter-Domain Routing). We've released this middleware in a
+Django package called [django-allow-cidr][]. The way it works is to store the normal hosts you've set
+in your `ALLOWED_HOSTS` setting, change that setting to `['*']` in order to bypass Django's default
+host header checking in as part of the `HttpRequest.get_host()` method, and do the checking itself.
+It does the checking via the same methods as Django would have, but if those methods fail it does
```
Comment (metadave) [inline _posts/2018-02-27-django-k8s-elb-health-checks.md]: command between `fail` and `it`?

File: _posts/2018-02-27-django-k8s-elb-health-checks.md (hunk 1 above)
Comment (pmac) [reply _posts/2018-02-27-django-k8s-elb-health-checks.md]: The sentence is just "...several of our SRE status reports." I've just linked the last 3 words.

Comment (pmac) [reply _posts/2018-02-27-django-k8s-elb-health-checks.md]: Ohh... nm.. you mean after "reports". Good catch. Done.

File: _posts/2018-02-27-django-k8s-elb-health-checks.md (hunk 6 above)
Comment (pmac) [reply _posts/2018-02-27-django-k8s-elb-health-checks.md]: Hmm... I'm not sure on that one.... doesn't seem to be a pause... I could be wrong.

**Comment by pmac** [comment]
Y'all don't think I was too revealing of our old `ALLOWED_HOSTS` ways? I think it's okay, but maybe I should do more to get across how wrong that is?

**Comment by pmac** [comment]
If not I'm happy to merge this if y'all are happy.

**Comment by metadave** [comment]
merge away

**Pull request merged**